Shift3/boilerplate-client-react
# feat(Unsaved Change prompt): adding prompts at navigation on forms

## Changes
1. Prompts before users navigate away from Forms
2. Adding the re-usable FormPrompt component so we can add the prompt to any forms we create in the future.

## Purpose
The purpose of this feature is to protect users from losing data from accidentally hitting the back button, and providing a better user experience.

## Approach
This change adds in prompt that only activates if you press any element that navigates the user away from their current page. This does not include the Submit buttons.

## Learning
I was able to accomplish the desired results by studying the following resources.
[isDirty - React Hook Form Docs](https://react-hook-form.com/api/useformstate)
[Prompt - React Router Docs](https://v5.reactrouter.com/core/api/Prompt)

## Testing
1. Pull in the changes to your local copy of this branch and restart Docker
2. Login to the application
3. Press "Add Agent"
4. Start filling out the form
5. Press any element that will navigate you away from the current page
6. _**ASSERT**_ Prompt appears stating you have unsaved changes
7. _**ASSERT**_ Once you select "Ok" you are navigated away from the current page
8. _**ASSERT**_ If you select "Cancel" you are redirected to the top of the current page without losing any form data.

`Follow this same process with the Change Password, User Profile, and Add Agent forms`

### Closes #496
